### ISSUE TITLE
Mario 64 Last impact slow downs

### ISSUE THREAD

**Issue by fzurita** [mupen64plus/mupen64plus-core#414]
I'm getting complains in the play store about Super Mario 64 last impact running significantly slower in the latest release. It's possible some recent change is causing a slow down somewhere. I did try the game and it does seem to run particularly slow.

**Comment by loganmc10** [comment]
The guy they wrote the ROM hack suggests Counter Factor = 1 in Project64

https://www.youtube.com/watch?v=1mqC4TBpxRU

It probably just needs CountPerOp=1, an entry could be added to the ini to default to that if that is the case

**Comment by fzurita** [comment]
I don't think that will help. It was more like the phone is struggling to keep up.

**Comment by fzurita** [comment]
So this apparently affects all Mario 64 hacks, I'm getting numerous reports of this in the Google play store. So I guess these hacks are important to people. I'll try to trace down which commit caused the slow downs.

**Comment by loganmc10** [comment]
Did you try CountPerOp=1? It can make a big difference in some cases.

**Comment by fzurita** [comment]
A lower CountPerOp setting will make the game harder to emulate, right? The problem is not that the emulator is not running fast enough, the problem is that the host CPU requirements are suddenly much higher to run the game smoothly.

**Comment by loganmc10** [comment]
> A lower CountPerOp setting will make the game harder to emulate, right? The problem is not that the emulator is not running fast enough, the problem is that the host CPU requirements are suddenly much higher to run the game smoothly.

The effects are different for each game. It's just a guess at that the timing should be for that particular game. If I get a chance I'll try to track down one of these ROM hacks and try it out. If you bisect it you'll probably come to some commit that changed the timing a bit, there was nothing done recently that would dramatically increase the load on the CPU, besides fiddling with the timing, which is why CountPerOp might be the solution

**Comment by fzurita** [comment]
Right, I ended up trying CountPerOp of 1 and it does indeed make these games run faster. Actually, for these Mario 64 hacks, the higher the CountPerOp, the slower the game runs. I guess I hadn't seen that before.

**Comment by fzurita** [comment]
I wonder if this is producing some odd result for Mario 64 hacks:

```
vi->count_per_scanline = ((vi->clock / vi->expected_refresh_rate) / (vi->regs[VI_V_SYNC_REG] + 1));
```

That was the change in https://github.com/mupen64plus/mupen64plus-core/commit/2379d07f85498517c6d91e39a2c4e7c2967172d9

**Comment by loganmc10** [comment]
Yeah that change was probably the one that did it. Resident Evil 2 didn't used to have voices when CountPerOp=1 for instance, but after that change, the voices work with CountPerOp=1 (just as an example of how it can change things)

A lot of ROM hacks seem to work better with CountPerOp=1 (Goldfinger 007 is another recent example).

Do they work worse now with CountPerOp=1 than they used to with CountPerOp=2? Or is it the same?

**Comment by fzurita** [comment]
I'll have to find an older version and compare.

**Comment by fzurita** [comment]
Ok, so I'm getting some very broken numbers for this calculation in Mario 64 hacks.

```
vi->count_per_scanline = ((vi->clock / vi->expected_refresh_rate) / (vi->regs[VI_V_SYNC_REG]
```

That above is giving me a value of ```386```.

**Comment by loganmc10** [comment]
Yeah it looks like it's writing 2100 to V_SYNC_REG which is definitely a bogus value. I'll figure out how to patch my Mario ROM and poke around, but honestly, these ROM hacks don't work on a real N64, their days are numbered, especially if we are going to strive for accuracy

**Comment by fzurita** [comment]
Right, then the pitchforks come after me :)

**Comment by loganmc10** [comment]
Oh actually that ROM was easier to find than I thought, I thought I'd have to patch a SM64 ROM.

Yeah, it writes 2100 to V_SYNC_REG. To give you an idea, Goldfinger 007, a really cool GoldenEye ROM hack that actually works on a real N64, writes 525 to V_SYNC_REG like a retail game.

A lot of these ROM hacks were designed and tested specifically for PJ64 back in the day. Maybe the 2100 was some kind of overclocking trick or whatnot, but it's kind of a tricky situation. I'd rather not put hacky stuff into the emulator to cater to poorly written ROM hacks, but other might disagree, we could potentially put in some kind of check for V_VSYNC_REG, because we know what the value should be:

```
VI_V_SYNC_REG controls the number of lines per frame (or the number of half-lines per field). Values are typically taken straight from the NTSC or PAL specifications. For NTSC, this is 0x20D (525 lines). For PAL, this is 0x271 (625 lines).
```
But I don't know exactly how all that works, I think it depends if the game is in interlaced mode or not, like Indiana Jones starts off writing 525 to that register, then changes to 524 when the graphics appear
